File: ios/RitualCircles/RitualCircles/Views/OnboardingView.swift
import SwiftUI

struct OnboardingView: View {
    @EnvironmentObject private var state: AppState

    @State private var userName: String = ""
    @State private var firstName: String = ""
    @State private var lastName: String = ""
    @State private var city: String = ""
    @State private var availabilityDay: String = "Mon"
    @State private var availabilityTime: String = "18:00:00"
    @State private var modality: Modality = .online
    @State private var ritualType: String = "coffee"

    @State private var inviteCode: String = ""
    @State private var isSubmitting: Bool = false
    @State private var error: String?

    private let days = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]

    var body: some View {
        VStack(spacing: 16) {
            Text("Set up your profile")
                .font(.headline)
                .frame(maxWidth: .infinity, alignment: .leading)

            VStack(spacing: 12) {
                TextField("Username (unique)", text: $userName)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
                    .textFieldStyle(.roundedBorder)
                TextField("First name", text: $firstName)
                    .textInputAutocapitalization(.words)
                    .textFieldStyle(.roundedBorder)
                TextField("Last name", text: $lastName)
                    .textInputAutocapitalization(.words)
                    .textFieldStyle(.roundedBorder)

                TextField("City (optional)", text: $city)
                    .textInputAutocapitalization(.words)
                    .textFieldStyle(.roundedBorder)

                HStack {
                    Picker("Day", selection: $availabilityDay) {
                        ForEach(days, id: \.self) { Text($0) }
                    }
                    .pickerStyle(.menu)

                    TextField("Time (HH:MM:SS)", text: $availabilityTime)
                        .keyboardType(.numbersAndPunctuation)
                        .textFieldStyle(.roundedBorder)
                }

                Picker("Modality", selection: $modality) {
                    Text("Online").tag(Modality.online)
                    Text("Offline").tag(Modality.offline)
                }
                .pickerStyle(.segmented)

                TextField("Ritual type (e.g. coffee)", text: $ritualType)
                    .textInputAutocapitalization(.never)
                    .textFieldStyle(.roundedBorder)
            }

            if let error {
                Text(error).foregroundStyle(.red).font(.footnote)
                    .frame(maxWidth: .infinity, alignment: .leading)
            }

            VStack(spacing: 10) {
                Button {
                    Task { await submitProfileOnly() }
                } label: {
                    Text(isSubmitting ? "Saving..." : "Save profile")
                        .frame(maxWidth: .infinity)
                }
                .buttonStyle(.borderedProminent)
                .disabled(isSubmitting || !profileReady)

                Button {
                    Task { await createCircle() }
                } label: {
                    Text("Create circle")
                        .frame(maxWidth: .infinity)
                }
                .buttonStyle(.bordered)
                .disabled(isSubmitting || !profileReady)

                HStack(spacing: 8) {
                    TextField("Invite code", text: $inviteCode)
                        .textInputAutocapitalization(.never)
                        .textFieldStyle(.roundedBorder)

                    Button("Join") {
                        Task { await joinCircle() }
                    }
                    .buttonStyle(.bordered)
                    .disabled(isSubmitting || inviteCode.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
                }
            }

            Spacer()
        }
        .padding()
    }

    private var profileReady: Bool {
        !userName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            && !firstName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            && !lastName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    private func submitProfileOnly() async {
        await patchMe()
        await state.refreshHome()
    }

    private func patchMe() async {
        isSubmitting = true
        error = nil
        defer { isSubmitting = false }

        do {
            _ = try await APIClient.shared.patchMe(
                UserUpdateRequest(
                    user_name: userName.trimmingCharacters(in: .whitespacesAndNewlines).lowercased(),
                    first_name: firstName.trimmingCharacters(in: .whitespacesAndNewlines),
                    last_name: lastName.trimmingCharacters(in: .whitespacesAndNewlines),
                    city: city.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : city,
                    availability_day: availabilityDay,
                    availability_time: availabilityTime
                )
            )
        } catch {
            self.error = error.localizedDescription
        }
    }

    private func createCircle() async {
        await patchMe()
        guard error == nil else { return }

        isSubmitting = true
        defer { isSubmitting = false }
        do {
            let circle = try await APIClient.shared.createCircle(
                CircleCreateRequest(
                    ritualType: ritualType.trimmingCharacters(in: .whitespacesAndNewlines),
                    modality: modality,
                    recurringTime: "\(availabilityDay) \(String(availabilityTime.prefix(5)))",
                    city: modality == .offline ? (city.isEmpty ? nil : city) : nil,
                    countryCode: nil,
                    cityName: nil,
                    meetingPlace: nil,
                    inviteOnly: true
                )
            )
            _ = circle
            await state.refreshHome()
        } catch {
            self.error = error.localizedDescription
        }
    }

    private func joinCircle() async {
        await patchMe()
        guard error == nil else { return }

        isSubmitting = true
        defer { isSubmitting = false }
        do {
            _ = try await APIClient.shared.joinCircle(inviteCode: inviteCode.trimmingCharacters(in: .whitespacesAndNewlines))
            await state.refreshHome()
        } catch {
            self.error = error.localizedDescription
        }
    }
}

File: android/app/src/main/java/com/ritualcircles/ui/views/OnboardingView.kt
package com.ritualcircles.ui.views

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.ritualcircles.models.CircleCreateRequest
import com.ritualcircles.models.UserUpdateRequest
import com.ritualcircles.ui.AppState
import kotlinx.coroutines.launch

@Composable
fun OnboardingView(state: AppState, onDone: () -> Unit) {
    var userName by remember { mutableStateOf("") }
    var firstName by remember { mutableStateOf("") }
    var lastName by remember { mutableStateOf("") }
    var city by remember { mutableStateOf("") }
    var day by remember { mutableStateOf("Mon") }
    var time by remember { mutableStateOf("18:00:00") }
    var modality by remember { mutableStateOf("online") }
    var ritualType by remember { mutableStateOf("coffee") }
    var inviteCode by remember { mutableStateOf("") }

    var working by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    val scope = rememberCoroutineScope()

    fun recurringTime(): String = "$day ${time.take(5)}"

    fun profileReady(): Boolean =
        userName.isNotBlank() && firstName.isNotBlank() && lastName.isNotBlank()

    fun profilePatch() = UserUpdateRequest(
        user_name = userName.trim().lowercase(),
        first_name = firstName.trim(),
        last_name = lastName.trim(),
        city = city.trim().ifBlank { null },
        availability_day = day,
        availability_time = time,
    )

    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Text("Onboarding", style = MaterialTheme.typography.titleMedium)
        Text("Set your profile, then create or join a circle.", style = MaterialTheme.typography.bodySmall)

        OutlinedTextField(value = userName, onValueChange = { userName = it }, label = { Text("Username (unique)") })
        OutlinedTextField(value = firstName, onValueChange = { firstName = it }, label = { Text("First name") })
        OutlinedTextField(value = lastName, onValueChange = { lastName = it }, label = { Text("Last name") })
        OutlinedTextField(value = city, onValueChange = { city = it }, label = { Text("City (optional)") })

        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            OutlinedTextField(modifier = Modifier.weight(1f), value = day, onValueChange = { day = it }, label = { Text("Day") })
            OutlinedTextField(modifier = Modifier.weight(1f), value = time, onValueChange = { time = it }, label = { Text("Time HH:MM:SS") })
        }

        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            OutlinedTextField(modifier = Modifier.weight(1f), value = modality, onValueChange = { modality = it }, label = { Text("Modality") })
            OutlinedTextField(modifier = Modifier.weight(1f), value = ritualType, onValueChange = { ritualType = it }, label = { Text("Ritual type") })
        }

        error?.let { Text(it, color = MaterialTheme.colorScheme.error) }

        Button(
            enabled = !working && profileReady(),
            onClick = {
                scope.launch {
                    working = true; error = null
                    try {
                        state.api().patchMe(profilePatch())
                        onDone()
                    } catch (t: Throwable) { error = t.message }
                    working = false
                }
            }
        ) { Text("Save profile") }

        OutlinedButton(
            enabled = !working && profileReady(),
            onClick = {
                scope.launch {
                    working = true; error = null
                    try {
                        state.api().patchMe(profilePatch())
                        state.api().createCircle(CircleCreateRequest(ritualType = ritualType.trim(), modality = modality, recurringTime = recurringTime(), city = if (modality == "offline") city.trim().ifBlank { null } else null))
                        onDone()
                    } catch (t: Throwable) { error = t.message }
                    working = false
                }
            }
        ) { Text("Create circle") }

        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            OutlinedTextField(modifier = Modifier.weight(1f), value = inviteCode, onValueChange = { inviteCode = it }, label = { Text("Invite code") })
            Button(
                enabled = !working && inviteCode.isNotBlank() && profileReady(),
                onClick = {
                    scope.launch {
                        working = true; error = null
                        try {
                            state.api().patchMe(profilePatch())
                            state.api().joinCircle(inviteCode.trim())
                            onDone()
                        } catch (t: Throwable) { error = t.message }
                        working = false
                    }
                }
            ) { Text("Join") }
        }
    }
}
